torch_xla/csrc/helpers.h

```diff
@@ -322,10 +322,13 @@ class XlaHelpers {
   static xla::XlaOp PromotedLogicalUnaryOp(
       xla::XlaOp op, const std::function<xla::XlaOp(xla::XlaOp)>& unary_op);
 
-  template <typename T>
-  static xla::Literal Range(T start, T end, T step) {
+  // T is the returned type, A is the type used for accumulation. In general,
+  // A should have higher-or-equal-precision to T.
+  template <typename T, typename A = T>
+  static xla::Literal Range(A start, A end, A step) {
+    std::vector<A> accumulated = runtime::util::Range<A>(start, end, step);
```

Comment:
Should `runtime::util::Range` just be defined in terms of double, or just written more safely? e.g. creating an iota and multiplying it by `step` should prevent accumulating imprecision and hangs.

This is an easy mistake to make, so I'd rather make `Range` less dangerous if we can.

Reply:
Using double for everything would have some corner cases with integral types I think, but I agree a better fix would be in `Range`. I'll see if I can find how the upstream is handling this.

Reply:
We could add one or two safeguards in the runtime::range impl, where we check if the step is >= 0, and/or cur and prev are not the same in each iteration. If we would support negative step, then we would need to update this path, too:
```
} else {
    for (; start > end; start += step) {
      result.push_back(start);
    }
  }
 ```
To be clear, the range here is inclusive? I thought the general range functions are exclusive (of end).

Reply:
It's exclusive, the 260 in the output comes from 259's bfloat16 representation:

```
>>> torch.Tensor([259]).to(dtype=torch.bfloat16)
tensor([260.], dtype=torch.bfloat16)
```

Will pointed me to this xla::Array method we might be able to use instead of having our own Range implementation, I'm going to give it a shot: https://github.com/openxla/xla/blob/0756a113f9e65ac46805080e3171b030e6a9b4c6/xla/array.h#L244-L249